File: src/main/java/de/kitsunealex/projectx/item/ItemBlockBase.scala
package de.kitsunealex.projectx.item

import java.util

import cpw.mods.fml.relauncher.{Side, SideOnly}
import de.kitsunealex.projectx.block.BlockBase
import net.minecraft.block.Block
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.item.{Item, ItemBlock, ItemStack}

class ItemBlockBase(block: Block) extends ItemBlock(block) {

  setHasSubtypes(block.asInstanceOf[BlockBase].getSubNames != null)

  override def getMetadata(damage : Int): Int = if(block.asInstanceOf[BlockBase].getSubNames != null) damage else super.getMetadata(damage)

  @SideOnly(Side.CLIENT)
  override def getSubItems(item: Item, tab: CreativeTabs, items: util.List[_]): Unit = block.getSubBlocks(item, tab, items)

  @SideOnly(Side.CLIENT)
  override def getUnlocalizedName(stack: ItemStack): String = {
    if(block.asInstanceOf[BlockBase].getSubNames != null) {
      return s"${getUnlocalizedName()}.${block.asInstanceOf[BlockBase].getSubNames(stack.getItemDamage)}"
    }
    else {
      return getUnlocalizedName()
    }
  }

}

File: src/main/java/de/kitsunealex/projectx/item/ItemBlockBase.java
/*
 * This file is part of ProjectX.
 * Copyright (c) 2015 - 2018, KitsuneAlex, All rights reserved.
 *
 * ProjectX is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * ProjectX is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with ProjectX.  If not, see <http://www.gnu.org/licenses/lgpl>.
 */

package de.kitsunealex.projectx.item;

import de.kitsunealex.projectx.util.ISubtypeHolder;
import net.minecraft.block.Block;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;
import net.minecraft.util.NonNullList;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.annotation.Nullable;
import java.util.List;

public class ItemBlockBase extends ItemBlock {

    public ItemBlockBase(Block block) {
        super(block);
        setHasSubtypes(block instanceof ISubtypeHolder);
    }

    @Override
    public int getMetadata(int damage) {
        return block instanceof ISubtypeHolder ? damage : super.getMetadata(damage);
    }

    @Override
    public void getSubItems(CreativeTabs tab, NonNullList<ItemStack> items) {
        if(tab == block.getCreativeTabToDisplayOn()) {
            block.getSubBlocks(tab, items);
        }
    }

    @Override
    @SideOnly(Side.CLIENT)
    public String getUnlocalizedName(ItemStack stack) {
        if(block instanceof ISubtypeHolder) {
            String[] subNames = ((ISubtypeHolder)block).getSubNames();
            return String.format("%s.%s", getUnlocalizedName(), subNames[stack.getMetadata()]);
        }
        else {
            return getUnlocalizedName();
        }
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack stack, @Nullable World world, List<String> tooltip, ITooltipFlag flag) {
        block.addInformation(stack, world, tooltip, flag);
    }

}
